sled-agent/src/sled_agent.rs

```diff
@@ -387,21 +387,33 @@ impl SledAgent {
             storage.zone_bundler().clone(),
         )?;
 
-        match config.vmm_reservoir_percentage {
-            Some(sz) if sz > 0 && sz < 100 => {
-                instances.set_reservoir_size(&hardware, sz).map_err(|e| {
-                    error!(log, "Failed to set VMM reservoir size: {e}");
-                    e
-                })?;
-            }
-            Some(0) => {
-                warn!(log, "Not using VMM reservoir (size 0 bytes requested)");
-            }
-            None => {
-                warn!(log, "Not using VMM reservoir");
+        // Configure the VMM reservoir as either a percentage of DRAM or as an
+        // exact size in MiB.
+        let reservoir_mode = match (
+            config.vmm_reservoir_percentage,
+            config.vmm_reservoir_size_mb,
+        ) {
+            (None, None) => ReservoirMode::None,
+            (Some(p), None) => ReservoirMode::Percentage(p),
+            (None, Some(mb)) => ReservoirMode::Size(mb),
+            (Some(_), Some(_)) => panic!(
+                "only one of vmm_reservoir_percentage and \
+                vmm_reservoir_size_mb is allowed"
+            ),
+        };
+
+        match reservoir_mode {
+            ReservoirMode::None => warn!(log, "Not using VMM reservoir"),
+            ReservoirMode::Size(0) | ReservoirMode::Percentage(0) => {
+                warn!(log, "Not using VMM reservoir (size 0 bytes requested)")
             }
-            Some(sz) => {
-                panic!("invalid requested VMM reservoir percentage: {}", sz);
+            _ => {
+                instances
+                    .set_reservoir_size(&hardware, reservoir_mode)
+                    .map_err(|e| {
```

Comment:
It would be nice to use the `tap` crate for situations like this:  [tap_err()](https://docs.rs/tap/latest/tap/trait.TapFallible.html#tymethod.tap_err)